lib/StackAnalysis/StackAnalysis.cpp

```diff
@@ -383,6 +682,1178 @@ void commitToModel(GeneratedCodeBasicInfo &GCBI,
+
+#ifdef NORETURN_IS_FIXED
+      if (Oracle.getFunctionType(PCCallee) != FunctionTypeValue::NoReturn)
+#else
+      if (true)
+#endif
+        Queue.insert(Succ);
+      continue;
+    }
+
+    for (auto *Succ : successors(Current)) {
+      if (!GCBI::isPartOfRootDispatcher(Succ))
+        Queue.insert(Succ);
+    }
+  }
+
+  ValueToValueMapTy VMap;
+  SmallVector<BasicBlock *, 8> BlocksToExtract;
+  AnyPCCloned = UnexpectedPCCloned = nullptr;
+
+  for (const auto &BB : BlocksToClone) {
+    auto *Cloned = CloneBasicBlock(BB, VMap, Twine("__cloned"), Root);
+
+    if (BB == GCBI->anyPC())
+      AnyPCCloned = Cloned;
+    else if (BB == GCBI->unexpectedPC())
+      UnexpectedPCCloned = Cloned;
+
+    VMap[BB] = Cloned;
+    BlocksToExtract.emplace_back(Cloned);
+  }
+
+  remapInstructionsInBlocks(BlocksToExtract, VMap);
+
+  // Fix successor when encountering a call-site and fix fall-through in
+  // presence of a noreturn function.
+  std::map<llvm::CallInst *, MetaAddress> CallMap;
+  for (const auto &BB : BlocksToExtract) {
+    if (isFunctionCall(BB)) {
+      auto *Term = BB->getTerminator();
+      auto *CI = getFunctionCall(Term);
+
+      // If the function callee is null, we are dealing with an indirect call
+      auto *Next = getFunctionCallCallee(Term);
+      auto PCCallee = Next ? getBasicBlockPC(Next) : MetaAddress::invalid();
+      auto CalleeType = Oracle.getFunctionType(PCCallee);
+
+      if (CalleeType != FunctionTypeValue::NoReturn) {
+        auto *Br = BranchInst::Create(getFallthrough(Term));
+        ReplaceInstWithInst(Term, Br);
+      } else if (CalleeType == FunctionTypeValue::NoReturn) {
+#ifdef NORETURN_IS_FIXED
+        auto *Abort = CallInst::Create(M.getFunction("abort"));
+        new UnreachableInst(Term->getContext(), BB);
+        ReplaceInstWithInst(Term, Abort);
+#else
+        auto *Br = BranchInst::Create(getFallthrough(Term));
+        ReplaceInstWithInst(Term, Br);
+#endif
+      }
+
+      // Fix function_call marker
+      unsigned ArgNo = 0;
+      auto *I8PtrTy = Type::getInt8PtrTy(M.getContext());
+      auto *I8NullPtr = ConstantPointerNull::get(I8PtrTy);
+      for (Value *Arg : CI->args()) {
+        if (isa<BlockAddress>(Arg)) {
+          CI->setArgOperand(ArgNo, I8NullPtr);
+          if (Arg->use_empty())
+            cast<BlockAddress>(Arg)->destroyConstant();
+        }
+        ++ArgNo;
+      }
+
+      CallMap.insert({ CI, PCCallee });
+    }
+  }
+
+  // Extract outlined function
+  CodeExtractorAnalysisCache CEAC(*Root);
+  Function *OutlinedFunction = CodeExtractor(BlocksToExtract)
+                                 .extractCodeRegion(CEAC);
+
+  revng_assert(OutlinedFunction != nullptr);
+  revng_assert(OutlinedFunction->arg_size() == 0);
+  revng_assert(OutlinedFunction->getReturnType()->isVoidTy());
+
+  auto It = OutlinedFunction->user_begin();
+  cast<Instruction>(*It)->getParent()->eraseFromParent();
```

Comment:
Who is this user? Document.
Is it the only one? Maybe you should assert there's only one user.